src/grab.ts

```diff
@@ -1,157 +1,121 @@
 import { z } from "zod";
-import { BaseResult, ValueOf } from "./types";
+import { ValueOf } from "./types";
+import { ArrayResult, BaseResult, EntityResult } from "./builder";
 
-/**
- * Create a "projection" to grab fields from a document/list of documents.
- * @param {object} selection - Fields to grab
- * @param {object} conditionalSelections – Conditional fields to grab
- */
-export const grab =
-  <
-    S extends Selection,
-    CondSelections extends Record<string, Selection> | undefined
-  >(
-    selection: S,
-    conditionalSelections?: CondSelections
-  ) =>
-  <T>(prev: BaseResult<T>) => {
-    type FromSelection<T extends Selection> = z.ZodObject<{
-      [K in keyof T as K extends `${infer Key}:${string}`
-        ? Key
-        : K]: T[K] extends BaseResult<any> ? T[K]["schema"] : FromField<T[K]>;
-    }>;
-    type KeysFromSelection<T extends Selection> =
-      (keyof T extends `${infer Key}:${string}` ? Key : T) & string;
+export const grab = <
+  T extends z.ZodTypeAny | z.ZodArray<any>,
+  S extends Selection,
+  CondSelections extends Record<string, Selection> | undefined
+>(
+  query: string,
+  schema: T,
+  selection: S,
+  conditionalSelections?: CondSelections
+) => {
+  type FromSelection<Sel extends Selection> = z.ZodObject<{
```

Comment:
This type could be moved down with your other utility types, as it doesn't have any dependencies inside this function.